[Merged by Bors] - feat: improvements to RefinedDiscrTree

JovanGerb opened the pull request:
This PR defines `RefinedDiscrTree`, which is an improved version of `DiscrTree` with many new features, such as
- indexing lambda terms, dependent forall terms, and bound variables.
- giving a score to matches, in order to sort them
- indexing star patterns so that `a+b` and `a+a` are indexed differently
- taking into account eta reductions, so that `exp` can still be matched with the library pattern `fun x => exp (f x)`.

This PR makes `RefinedDiscrTree` into a lazy data structure, meaning that it can be used without a cache, just like `LazyDiscrTree`.

This PR also removes these features:
- indexing `fun x => x` as `id`
- indexing `fun x => f x + g x` as `f + g`, and similarly for `-`, `*`, `/`, `⁻¹`, `^`.
- indexing `fun _ => 42` as `42`

These equivalent indexings do not hold definitionally in the `reducible` transparency, which is the transparency that is used for unification when using a discrimination tree. Therefore, indexing these different expressions the same is actually inefficient rather than helpful.

This is part of the bigger #11768, which uses this discrimination tree for a library rewriting tactic.

This replaces an older version of `RefinedDiscrTree`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

JovanGerb commented:
awaiting-review

rosborn commented on Mathlib/Lean/Meta/RefinedDiscrTree.lean:
```diff
@@ -0,0 +1,1204 @@
+partial def skipEntries (t : Trie α) (skipped : Array Key) : Nat → M (Array Key × Trie α)
+  | 0      => pure (skipped, t)
+  | skip+1 =>
+    t.children!.foldr (init := failure) fun (k, c) x =>
```
Folding with an `Alternative` seems like it should be abstracted, but I guess lean doesn't have `Foldable` types?

rosborn replied:
Ok. I see that `List` has `firstM`. This is essentially `lastM` but for `Array`?

JovanGerb replied:
I would also prefer if this fold with an alternative could be abstracted. I wasn't aware of `List.firstM`. That is essentially what I would need here, but for `Array`. Having the `Foldable` type class in Lean would solve this :)

rosborn commented on Mathlib/Lean/Meta/RefinedDiscrTree.lean:
```diff
@@ -0,0 +1,1204 @@
+  let result ← go e 0
+  return result.qsort (·.2.1 > ·.2.1)
+where
+  /-- Auxiliary function for `getMatchWithScoreWithExtra` -/
```
```suggestion
```

rosborn commented on Mathlib/Lean/Meta/RefinedDiscrTree.lean:
```diff
@@ -0,0 +1,1204 @@
+    k
+  else do
+    let e ← withReader (fun c => { c with bvars := lambdas ++ c.bvars }) k
+    return lambdas.foldl (fun _ => ·.lam) e
```
```suggestion
    return lambdas.foldl (fun d _ => d.lam) e
```
Don't like the use of `·` with an anonymous `fun`.

rosborn reviewed (approved):
Overall looks good. I would definitely want a second pair of eyes, as I haven't reviewed this style of code in a while.
I also wonder if `isNumeral` and `reduce` are useful outside of this file, but I'm not familiar enough with Mathlib's metaprogramming files to comment.

JovanGerb commented:
> I also wonder if `isNumeral` and `reduce` are useful outside of this file

These functions are defined in the same exact way in Lean's `DiscrTree`. So If they were useful anywhere else, they should be used from there.

JovanGerb commented on Mathlib/Lean/Meta/RefinedDiscrTree.lean:
```diff
@@ -0,0 +1,1204 @@
+      distributeLambdas lambdas type lhs rhs
+    | [] => return (type, lhs, rhs, [])
+
+/-- Normalize an application if the head is  `+`, `*`, `-` or `/`.
```
I think so. In those we have the rules `fun i => s • x i` = `s • x` and `fun i => x i ^ b` = `x ^ b`. I think I'll move this part of the code to a separate file (the part for normalizing pi instances). In the long run when `RefinedDiscrTree` will move to std/Boost, this bit of code needs to stay in Mathlib because the pi instances are defined here. Then there should be a way to dynamically extend the reduction function. Then this code could go into `Mathlib.Algebra.Group.Pi.basic`, do it would run this if and only if the pi instances are imported.

JovanGerb replied:
I have reimplemented this code now, adding support for HSMul, HVAdd and HPow. In the process I also discovered a bug in an edge case. I have now added a test file to see that it is indeed doing what it should. There is now the function `reducePi` which does this normalization returning the result as an `Expr`, so that it can also be used somewhere else if that is useful.

joneugster commented:
It looks like [Mathlib.Tactic.FunProp.RefinedDiscrTree](https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Tactic/FunProp/RefinedDiscrTree.lean) (authored by yourself) is already in mathlib and almost identical to the main file added here.

Could you please update this PR so that it either updates the existing file, or replaces the usage of that in favour of your new file?

JovanGerb commented:
As mentioned on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Point.26Click.20library.20rewrite.20tactic/near/447023540), `RefinedDiscrTree` should be made lazy, so that it doesn't require cache.

As for [Mathlib.Tactic.FunProp.RefinedDiscrTree](https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Tactic/FunProp/RefinedDiscrTree.lean), it would probably be good to update it to the current most recent version of `RefinedDiscrTree`, but eventually it should switch to the lazy variant when I make it.

JovanGerb commented:
I have rewritten the `RefinedDiscrTree` implementation, based on `LazyDiscrTree`. It is now a lazy data structure, meaning that it can be used for my library rewrite tactic without any pre built cache. I've also removed the old `RefinedDiscrTree` implementation from `fun_prop`, and replaced it with the new implementation (cc @lecopivo).

awaiting-review

JovanGerb commented:
awaiting-review

JovanGerb commented:
For plugging the lazy discrimination tree into `fun_prop`, the environment now contains the lazy discrimination tree instead of a strict one, which is good because it means we can avoid inserting lemmas with very long key sequences when we don't need them. However I wasn't sure how to replace the `RefinedDiscrTree` in the environment with an updated one (with some more nodes evaluated). So what I did instead is to store the progress of evaluating the lazy tree in the state of a single `fun_prop` call, and then throw this progress away at the end.

Another thing is that the data structure is not persistent: the array of all nodes needs to be copied from the environment for every `fun_prop` call

Both of these aren't a performance problem though, because evaluating the tree is quite fast, and the array of nodes in an un-evaluated tree is pretty short in the case of `fun_prop`.

joneugster commented:
Thanks for updating this! I might only be able to look at it early next week.

I removed the "WIP" label as this prevents reviwers from seeing the PR on the queue

eric-wieser commented on Mathlib/Tactic/FunProp/RefinedDiscrTree.lean:
Can you explain in the PR description what happened here? Did you split the file? Rewrite it from scratch? Move it with just enough modifications that Lean can no longer track it?

eric-wieser replied:
I think the diff might be easier to process if you make an initial PR that just moves `Tactic.FunProp.RefinedDiscrTree` to `Lean.Meta.RefinedDiscrTree`, without any other modifications.

That will at least cause github to show a diff in the long docstring, rather than showing it entirely added then deleted.

JovanGerb replied:
I split the one file into multiple files, and the files changed quite a lot in the process of making the disctimination tree lazy, so I think there won't be much of a meaningful diff.

eric-wieser replied:
I split out the move, and have resolved the superficial conflicts

eric-wieser replied:
I've now created #20381 to contain the split, without any refactoring. I'll resolve the merge conflicts by overwriting them with the contents of the current PR, so the final code will be unchanged; but hopefully this will cut the diff by 30%.

JovanGerb replied:
It seems there is still a merge conflict left over. How can I get rid of it?

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Pi.lean:
```diff
@@ -0,0 +1,269 @@
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
+  if h : args.size < 5 then return none else
+  let rec instH := match n with
+    | `HVAdd.hVAdd => `instHVAdd
+    | `HSMul.hSMul => `instHSMul
```
Can you use ``` `` ``` to make this an error if the names change?

JovanGerb replied:
I did that oroginally, but it requires to import more, and then it complains about an unnecessary import

eric-wieser replied:
The unnecessary import can be silenced by running `lake exe shake --update`. Unless the import is particularly expensive (which CI will tell you), I think robustness matters more.

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Pi.lean:
```diff
@@ -0,0 +1,269 @@
+  etaExpand args type lambdas 6 fun args lambdas _ => do
+    let lhs := args[4]
+    let rhs := args[5]
+    reduceAux type inst [lhs, rhs] lambdas args #[inst] 6 2
+      (lambdaBody? 3 >=> app6? n >=> fun (inst, l, r) =>
+        -- check for the form `fun a b x => a x + b x`
```
Is this cheaper than doing an `isDefEq` check against the canonical instances, as would happen with `~q()` matching? It certainly seems more fragile this way.

JovanGerb replied:
It is cheaper indeed, but this code won't run too many times, so it's probably fine to use isDefEq instead.

JovanGerb replied:
What's going on here is checking whether the expression is of the form `fun a b x => a x + b x` and if so, returning the `Add` instance. To do this with `isDefEq` would mean creating a term with a bunch of fresh metavariables, and extracting the value of the instance metaviable. I feel like this would be more cumbersome.

JovanGerb replied:
Moreover, I want to check against the form `fun a b x => a x + b x`, not the canonical Pi instance, because there can also be other instances that reduce to `fun a b x => a x + b x`, and I also want to treat these as Pi instances.

eric-wieser replied:
Ah sorry, I misread your implementation and thought you were checking against the canonical Pi instance. I now see that the names are references to the non-heterogenous instances!

A defeq check against the pi instance would work in the way you want, but indeed is more complex without benefit.

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Initialize.lean:
```diff
@@ -0,0 +1,296 @@
+    (act : Name → ConstantInfo → MetaM (List (α × List (Key × LazyEntry))))
+    (tree : PreDiscrTree α) (name : Name) (constInfo : ConstantInfo) :
+    BaseIO (PreDiscrTree α) := do
+  if constInfo.isUnsafe then pure tree else
+  if LazyDiscrTree.blacklistInsertion env name then pure tree else
```
style nit:
```suggestion
  if constInfo.isUnsafe then return tree
  if LazyDiscrTree.blacklistInsertion env name then return tree
```

JovanGerb replied:
This code is performance critical since it gets run on most constants in the environment (when building the tree the first time). And the code doen't get compiled (instead it is run by the interpreter). When writing `if .. then return ..`, the compiler creates a separate function for the (implicit) else branch of the if statement, which is less efficient than a basic if-else.

eric-wieser replied:
Could you add a comment in the code to this effect?

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Lookup.lean:
```diff
@@ -0,0 +1,293 @@
+        let entry := (entry, value)
+        match star? key with
+        | some id =>
+          match stars[id]? with
+          | some idx => addLazyEntryToTrie idx entry
+          | none     => stars := stars.insert id (← newTrie entry)
```
optional: I find these easier to read as
```suggestion
          if let .some idx := stars[id]? then
            addLazyEntryToTrie idx entry
          else
            stars := stars.insert id (← newTrie entry)
```

JovanGerb commented:
Thanks for the review, but I'll probably only have time to work on it after next week.

eric-wieser commented on MathlibTest/RefinedDiscrTree.lean:
```diff
@@ -108,8 +108,13 @@ info: @Nat.fold ℕ 10 (@HAdd.hAdd (ℕ → (@LT.lt ℕ *0 #0 10) → ℕ → 
+open ENNReal in
+/-- info: @Eq ℝ (@HAdd.hAdd ℝ ℝ *0 *1 10 20) 0 -/
+#guard_msgs in
+# ((show PiLp 1 fun _ : Unit => ℝ from fun _ => 10) + (show PiLp 1 fun _ : Unit => ℝ from fun _ => 20)) () = 0
```
This should use `WithLp.Equiv` to do the casting, not `show`.

JovanGerb commented on Mathlib/Lean/Meta/RefinedDiscrTree.lean:
```diff
@@ -1,223 +1,150 @@
 
   For example, this allows for more specific matching with the left hand side of
-  `∑ i ∈ range n, i = n * (n - 1) / 2`, which is indexed by
+  `∑ i in range n, i = n * (n - 1) / 2`, which is indexed by
```
Yes, you're right

eric-wieser commented on MathlibTest/RefinedDiscrTree.lean:
```diff
@@ -109,10 +109,11 @@ info: @Nat.fold ℕ 10 (@HAdd.hAdd (ℕ → (@LT.lt ℕ *0 #0 10) → ℕ → 
-open ENNReal in
-/-- info: @Eq ℝ (@HAdd.hAdd ℝ ℝ *0 *1 10 20) 0 -/
+/--
+info: @Eq ℝ (@HAdd.hAdd ℝ ℝ *0 *1 (@DFunLike.coe (Equiv (PUnit.{u} → ℝ) (WithLp 1 (PUnit.{u} → ℝ))) *2 *3 *4 (@Equiv.symm (WithLp 1 (PUnit.{u} → ℝ)) (PUnit.{u} → ℝ) (WithLp.equiv 1 (PUnit.{u} → ℝ))) 10 PUnit.unit) (@DFunLike.coe (Equiv (PUnit.{u} → ℝ) (WithLp 1 (PUnit.{u} → ℝ))) *5 *6 *7 (@Equiv.symm (WithLp 1 (PUnit.{u} → ℝ)) (PUnit.{u} → ℝ) (WithLp.equiv 1 (PUnit.{u} → ℝ))) 20 PUnit.unit)) 0
```
I guess to make this less of a monster you could add an `def testEquiv := (WithLp.equiv 1 (Unit → ℝ))` on the line before

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Pi.lean:
```diff
@@ -1,115 +1,269 @@
+
+/-- Normalize an application if the head is `⁻¹` or `-`. -/
+private def reduceUnOp (n : Name) (args : Array Expr) (lambdas : List FVarId) :
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
```
`Option (List (Array (Option Expr) × List FVarId` appears 4 times in this file; can you perhaps replace it with a `structure` so that you can document what the pieces mean?

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Pi.lean:
```diff
@@ -1,115 +1,269 @@
+    MetaM (Option (Name × List (Array (Option Expr) × List FVarId))) := do
+  if let .const n _ := e.getAppFn then
+    match n with
+    | ``Neg.neg
+    | ``Inv.inv     => Option.map (n, ·) <$> reduceUnOp   n e.getAppArgs lambdas
+    | ``HVAdd.hVAdd
+    | ``HSMul.hSMul => Option.map (n, ·) <$> reduceHActOp n e.getAppArgs lambdas
+    | ``HPow.hPow   => Option.map (n, ·) <$> reduceHPow     e.getAppArgs lambdas
+    | ``HAdd.hAdd
+    | ``HMul.hMul
+    | ``HSub.hSub
+    | ``HDiv.hDiv   => Option.map (n, ·) <$> reduceHBinOp n e.getAppArgs lambdas
```
I think its likely we will want `inf` and `sup` here too (and to a lesser extent, all the other lattice operators). Can you either add an extension mechanism for adding more entries here, or a TODO comment about doing so?

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Pi.lean:
```diff
@@ -1,115 +1,269 @@
+        | type, [arg] => #[type, none, arg]
+        | _, _ => unreachable! )
+
+/-- Normalize an application if the head is `•` or `+ᵥ`. -/
+private def reduceHActOp (n : Name) (args : Array Expr) (lambdas : List FVarId) :
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
+  if h : args.size < 5 then return none else
+  let rec instH := match n with
+    | ``HVAdd.hVAdd => ``instHVAdd
+    | ``HSMul.hSMul => ``instHSMul
+    | _ => unreachable!
+  let some (α, type, inst) := args[3].app3? instH | return none
+  etaExpand args type lambdas 6 fun args lambdas _ => do
+    let a := args[4]
+    let arg := args[5]
+    reduceAux type inst [arg] lambdas args #[inst, a] 6 2
+      (lambdaBody? 3 >=> app6? n >=> fun (inst, a, b) =>
+        -- check for `fun a b x => a • b x` or `fun a b x => a +ᵥ b x`
+        if (a == .bvar 2) && b == .app (.bvar 1) (.bvar 0) then inst else none)
+      (fun
+        | type, [arg] => #[α, type, none, none, a, arg]
+        | _, _ => unreachable!)
+
+/-- Normalize an application if the head is `^`. -/
+private def reduceHPow (args : Array Expr) (lambdas : List FVarId) :
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
+  if h : args.size < 6 then return none else
+  let some (type, β, inst) := args[3].app3? ``instHPow | return none
+  let arg := args[4]
+  let exp := args[5]
+  reduceAux type inst [arg] lambdas args #[inst, exp] 6 2
+      (lambdaBody? 3 >=> app6? ``HPow.hPow >=> fun (inst, a, b) =>
+        -- check for `fun a b x => a x ^ b`
+        if (a == .app (.bvar 2) (.bvar 0) && b == .bvar 1) then inst else none)
+    (fun
+      | type, [arg] => #[type, β, none, none, arg, exp]
+      | _, _ => unreachable!)
```
Can these be merged by letting the caller pass in the argument positions?

JovanGerb commented:
The comments are about RefinedDiscrTree.Pi, which is the code that allows `fun x => f x + g x` and `f + g` to be indexed the same. However, this is problematic because these expressions aren't definitionally equal in the reducible transparency. For example in `simp` this would be useless.

I asked Tomas Skrivan, and he said that he doesn't need/use this feature in `fun_prop`.

So, I'll delete this part of the code. Also because the implementation is a bit awkward. If we ever need it, it can be added then.

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree.lean:
```diff
@@ -1,223 +1,131 @@
+    (ext : EnvExtension (IO.Ref (Option (RefinedDiscrTree α))))
+    (addEntry : Name → ConstantInfo → MetaM (List (α × List (Key × LazyEntry)))) (ty : Expr)
+    (constantsPerTask : Nat := 1000) (capacityPerTask : Nat := 128) : MetaM (MatchResult α) := do
+  let cctx ← (read : CoreM Core.Context)
```
```suggestion
  let cctx ← readThe Core.Context
```

eric-wieser commented:
> So, I'll delete this part of the code.

Please update the PR description to reflect this.

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Basic.lean:
```diff
@@ -1,178 +1,266 @@
+    | _ => return key.format
+  /-- Add parentheses if `paren == true`. -/
+  parenthesize (msg : MessageData) (paren : Bool) : MessageData :=
+    if paren then m! "({msg})" else msg
```
```suggestion
    if paren then msg.paren else msg
```
which knows about line wrapping. You might want to use `.group` in the else branch for line-wrapping reasons too.

eric-wieser replied:
I've upstreamed this to https://github.com/leanprover/lean4/pull/7200

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Basic.lean:
```diff
@@ -1,178 +1,266 @@
+    if nargs == 0 then
+      return f
+    else do
+      let mut r := f
+      for _ in [:nargs] do
+        r := r ++ m!" {← go}"
+      return parenthesize m!"{r}" paren
```
```suggestion
      let args ← (List.range nargs).mapM fun _ => go
      return parenthesize (MessageData.joinSep (f :: args) " ") paren
```
though you might want to consider using `Format.line` instead of `" "`

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Basic.lean:
```diff
@@ -1,178 +1,266 @@
+instance : Inhabited (LazyEntry) where
+  default := { mctx := {} }
 
+private def LazyEntry.format (entry : LazyEntry) : Format :=
+  let results := if entry.results matches [] then f!"" else f!", results: {entry.results}"
+  let todo    := if let some info := entry.previous then f!", todo: {info.expr}" else ""
+  f!"stack: {entry.stack}{results}{todo}"
```
nit: perhaps more readable as
```suggestion
private def LazyEntry.format (entry : LazyEntry) : Format := Id.run do
  let mut parts := #[f!"stack: {entry.stack}"]
  unless entry.results matches [] do
    parts := parts.push f!"results: {entry.results}"
  if let some info := entry.previous then
    parts := parts.push f!"todo: {info.expr}"
  return Format.joinSep parts.toList ", "
```

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Basic.lean:
```diff
@@ -1,178 +1,266 @@
+  let lines := tree.root.fold (init := #[]) fun lines key trie =>
+    lines.push (Format.nest 2 f! "{key} =>{Format.line}{go trie}")
+  if lines.size = 0 then
+    f! "<empty discrimination tree>"
```
I think the prevailing style is to have no space here.
```suggestion
    f!"<empty discrimination tree>"
```
(and many others throughout)

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Basic.lean:
```diff
@@ -1,178 +1,266 @@
+  if lines.size = 0 then
+    f! "<empty discrimination tree>"
+  else
+    lines.foldl (init := "Discrimination tree flowchart:") (· ++ Format.line ++ ·)
```
I think this is
```suggestion
    Format.line.joinsep (f!"Discrimination tree flowchart:" :: lines)
```
note also that `Format.line` does not actually *require* a line break; if you want to prevent this appearing on one line if `lines` is short, use `\n`.

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Encode.lean:
```diff
@@ -183,326 +74,372 @@ where
+private def mkNewStar : StateT LazyEntry MetaM Key :=
+  modifyGet fun entry => (.star entry.nStars, { entry with nStars := entry.nStars + 1 })
+
+@[inline] private def withLams (lambdas : List FVarId) (key : Key) : StateT LazyEntry MetaM Key :=
+  do match lambdas with
```
nit:
```suggestion
@[inline] private def withLams (lambdas : List FVarId) (key : Key) : StateT LazyEntry MetaM Key := do
  match lambdas with
```
I think is the usual style

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Encode.lean:
```diff
@@ -183,326 +74,372 @@ where
+    if isStarWithArg (.fvar fvarId) a && !f.getAppFn.isMVar then
+      match entry.cache.find? original with
+      | some (key :: keys) => return [(key, { entry with results := keys })]
+      | some [] => unreachable!
```
Could you add a comment explaining why this is unreachable? Or perhaps `panic! "the reason"` if you're not actually certain?

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Encode.lean:
```diff
@@ -183,326 +74,372 @@ where
+      MetaM (Array (Array Key)) := do
+    if todo.isEmpty then
+      return result
+    else
+      let (keys, entry) := todo.back!
+      let todo := todo.pop
+      match ← evalLazyEntry entry with
+      | some xs =>
+        let rec /--
+          This variation on `List.fold` ensures that the array `keys`
+          isn't copied unnecessarily. -/
+        fold xs todo :=
+          match xs with
+          | [] => todo
+          | (key, entry) :: [] => todo.push (keys.push key, entry)
+          | (key, entry) :: xs => fold xs (todo.push (keys.push key, entry))
+        go (fold xs todo) result
+      | none =>
+        go todo (result.push keys)
```
nit: could unindent these lines and drop the `else`

JovanGerb replied:
It's a bit unfortunate that this nit causes the function to not be tail-recursive, and hence less efficient.

JovanGerb replied:
I'll add a comment about it

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Initialize.lean:
```diff
@@ -0,0 +1,298 @@
+the library search tree.
+-/
+private structure InitResults (α : Type) where
+  tree  : PreDiscrTree α := {}
+  errors : Array ImportFailure := #[]
```
nit: the spacing here (and in other structures) is a little weird; I think the convention is to just use one space around `:`, but I guess column aligning is ok if you feel strongly. Right now, this is neither!

eric-wieser commented on MathlibTest/RefinedDiscrTree.lean:
```diff
@@ -0,0 +1,171 @@
+
+open Qq Lean Meta RefinedDiscrTree
+
+macro "#" e:term : command =>
```
Could you use something less ambiguous like `#log_keys` or `#keys` here?

eric-wieser reviewed (commented):
It's hard for me to give this more than a superficial review I'm afraid; what follows are mainly style points and some suggestions around `MessageData`. Once you've addressed those, it is probably worth pinging on Zulip to try and find any other interested reviewers, or to get consensus that we're ok with a bus factor of 1 for this code!

eric-wieser commented on Mathlib/Lean/Meta/RefinedDiscrTree/Basic.lean:
```diff
@@ -1,178 +1,266 @@
+    if h : lines.size = 0 then
+      f! "<empty node>"
+    else
+      lines.foldl (init := lines[0]) (· ++ Format.line ++ ·) (start := 1)
```
This is another `Format.joinSep`

eric-wieser commented:
> Please update the PR description to reflect this.

To be clear: you should point out that this feature is being removed, not just that you are no longer adding it. The old version had some support for it, the new version does not.

JovanGerb commented:
Ooh interesting, merging master has revealed a bug in the matching procedure.

0art0 commented on Mathlib/Lean/Meta/RefinedDiscrTree/Basic.lean:
@JovanGerb explained to me that there is a trade-off between using persistent and ordinary data-structures in the implementation of (refined) discrimination trees.

On one hand, persistent data-structures will be better for tactics like `simp` that rely on a large list of tagged lemmas, since modified copies of the discrimination tree must be stored throughout the file in the environment extension. On the other hand, ordinary data-structures are better suited for library look-up tasks (like `rw??` and `exact?`), since there is less of an overhead for searching for entries in these data-structures.

The `RefinedDiscrTree` data-structure currently uses ordinary data-structures, and it may be pre-mature to switch to persistent data-structures immediately in the absence of a strong use-case. However, we agree that it would be good for this distinction to be documented somewhere (probably in the main doc-string at the top of the file) for future reference.

fpvandoorn replied:
I agree

fpvandoorn replied:
The explanation in the main file seems sufficient for this.

JovanGerb commented:
I fixed the above mentioned bug now, by refactoring the constructor `Key.star (id : Nat)` into `Key.star` and `Key.labelledStar (id : Nat)`.

Some potential todos:
- Add tests for the lookup algorithm, not just for the encoding algorithm.
- Add performance tests for building a large `RefinedDiscrTree`.

The reason why I didn't add these so far is that it seems more natural to have these tests tied to a tactic like `rw??`. So it's not clear to me that these should be added here.

JovanGerb commented:
!bench

grunweg commented:
Coming here from PR triage - let me just note that this PR has a merge conflict, so won't show up on the #queue right now.

JovanGerb commented:
I ran into an issue last week with `RefinedDiscrTree` when merging master (commit [1691adc](https://github.com/leanprover-community/mathlib4/pull/11968/commits/1691adc1fa7d5d796d37fea0a1b2d33585ab17a3)):
```
 error: ././././Mathlib/Probability/ProductMeasure.lean:89:44: `fun_prop` was unable to prove `Measurable ⇑(piCongrLeft X e)`
```
What happened here was that the target expression was not type correct in reducible transparency, since a type argument appeared in two different forms. In particular, in one there was an application of an `Equiv` using `DFunLike.coe`, and in the other an application of the underlying `toFun` (`DFunLike.coe` doesn't unfold in reducible transparency). Since the relevant theorem required the implicit argument to be the same in the two places, `RefinedDiscrTree` did not suggest it. That case could be solved using a dsimp, but it would be nicer if that wasn't necessary.

- This could be fixed by turning off the feature of labelled stars. This would be fine for `fun_prop`, but not a reasonable option for `rw??`, so I'd rather not do this.
- The way I currently fix this is by reducing in the `default` transparency inside implicit arguments. Actually `reducible_and_instances` would be sufficient to solve that particular problem.
  
  In principle, it's not good to unfold with a stronger transparency, but I couldn't think of any reasonable examples where this would stop a match from appearing.
- There is probably a way to deal with this more principally. Either to not index implicit arguments in the discrimination tree and instead indexing types of terms directly somehow. Or to use `isDefEq` unification on implicit arguments during lookup.

So anyways, I think it's probably fine like this.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

fpvandoorn commented on Mathlib/Lean/Meta/RefinedDiscrTree/Lookup.lean:
```diff
@@ -1,187 +1,278 @@
-/-- Return the possible `Trie α` that match with `n` metavariable. -/
-partial def skipEntries (t : Trie α) (skipped : Array Key) : Nat → M (Array Key × Trie α)
-  | 0      => pure (skipped, t)
+Currently, `.star` and `.labelledStar _` from the query expression are treated the same.
```
this should read: "`labelledStar` is unused", right?

JovanGerb replied:
You're right, this part was outdated.

fpvandoorn reviewed (approved):
LGTM

JovanGerb commented:
The test failed as expected. This is because I removed the cache-feature in encoding the expressions
